src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs

```diff
@@ -222,6 +222,25 @@ private void ValidateProcessedBlock(Block suggestedBlock, ProcessingOptions opti
     {
         IReleaseSpec spec = _specProvider.GetSpec(block.Header);
 
+        if (spec.BeaconStateRootAvailable && block.Header.BeaconStateRoot is Keccak beaconStateRoot)
+        {
+            if (!_stateProvider.AccountExists(BeaconStateRootPrecompile.Address))
+            {
+                _stateProvider.CreateAccount(BeaconStateRootPrecompile.Address, 1);
+            }
+
+            UInt256 timestamp = block.Header.Timestamp;
+
+            UInt256.Mod(timestamp, BeaconStateRootPrecompile.HISTORICAL_ROOTS_LENGTH, out UInt256 timestampReduced);
+            UInt256 rootIndex = timestampReduced + BeaconStateRootPrecompile.HISTORICAL_ROOTS_LENGTH;
+
+            StorageCell tsStorageCell = new(BeaconStateRootPrecompile.Address, timestampReduced);
+            StorageCell brStorageCell = new(BeaconStateRootPrecompile.Address, rootIndex);
+
+            _stateProvider.Set(tsStorageCell, timestamp.ToBigEndian());
+            _stateProvider.Set(brStorageCell, beaconStateRoot.Bytes.ToArray());
+        }
+
```

Comment:
This should be encapsulated on a few levels:
1. Part of the logic should be moved to `BeaconStateRootPrecompile`. Either the logic of calculating storage cells or even setting them in the state provider. This can be static method there.
2. Rest of logic should be encapsulated in either: a method in `BlockProcessor` or `BeaconRootManager`/`BeaconRootUpdater` (or better name) that should be called from here, similar as we do with `_blockTransactionsExecutor` or `_withdrawalProcessor`